### PR TITLE
[PackageBuilder] Fix test target ref in manifest for external packages

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This was caused due to test targets not being constructed when a package
acts as a dependency.

<rdar://problem/28038874> SR-2353: Test dependencies block a package
for being used downstream

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@ddunbar Should I write add functional test for this? not sure if this can be unit tested

**Comment by ddunbar** [comment]
I think this is important enough we need to be able to test it. What would it take to be able to mock up a PackageGraph enough to test this?

File: Sources/PackageLoading/PackageBuilder.swift [hunk 1]
```diff
@@ -208,15 +208,13 @@ public struct PackageBuilder {
...
-        let testModules: [Module]
-        if includingTestModules {
-            testModules = try constructTestModules(modules: modules)
-        } else {
-            testModules = []
-        }
+        // It is unfortunate that test modules has to be constructed even when they are
+        // not requested by the clients but they might be referenced in the manifest file
+        // so that needs to be validated.
+        let testModules = try constructTestModules(modules: modules)
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageBuilder.swift]: I think it is ok that we construct test modules here, the performance cost should be negligible and I think it is actually better to not have the graph structure change depending on the use case.

File: Sources/PackageLoading/PackageBuilder.swift [hunk 2]
```diff
@@ -208,15 +208,13 @@ public struct PackageBuilder {
...
         try fillDependencies(modules: modules + testModules)
-        let products = try constructProducts(modules, testModules: testModules)
-        return Package(manifest: manifest, path: packagePath, modules: modules, testModules: testModules, products: products)
+        let products = try constructProducts(modules, testModules: includingTestModules ? testModules : [])
+        return Package(manifest: manifest, path: packagePath, modules: modules, testModules: includingTestModules ? testModules : [], products: products)
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageBuilder.swift]: Are you sure this won't cause us to start building all tests for all packages by default?

Comment (aciidb0mb3r) [inline Sources/PackageLoading/PackageBuilder.swift]: Yeah it seems weird here, I'll add a FIXME

**Comment by aciidb0mb3r** [comment]
@ddunbar Added a mock package graph tester

**Comment by ddunbar** [comment]
This is great, thanks! LGTM

We should use this to check the cross package deps too... I wonder how close we are getting to being able to drop most of the functional tests. I guess we need testing for Build first.

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
I think it should be easily possible to run PackageBuilderTester on this since PackageBuilder is already run by the loader. I'll file a bug for investigating and start migrating cross package deps to unit testing

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
LGTM

**Comment by aciidb0mb3r** [comment]
Tested locally

**Pull request merged**